src/Microsoft.Health.Fhir.SqlServer/Features/Search/CustomQueries.cs

```diff
@@ -0,0 +1,55 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Data.SqlClient;
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Core;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Search
+{
+    internal static class CustomQueries
+    {
+        private static DateTimeOffset _lastUpdatedQueryCache = DateTimeOffset.MinValue;
+        public static readonly Dictionary<string, string> QueryStore = new Dictionary<string, string>();
+
+        public static string CheckQueryHash(SqlConnection connection, string hash, ILogger<SqlServerSearchService> logger)
+        {
+            var now = Clock.UtcNow;
+            if (now > _lastUpdatedQueryCache.AddMinutes(1))
```

Comment:
>if (now > _lastUpdatedQueryCache.AddMinutes(1)) [](http://example.com/codeflow?start=12&length=47)

We need to lock on entire execution to avoid dup calls to the database.